lib/runtime/src/pipeline/network/egress/addressed_router.rs

```diff
@@ -148,11 +152,29 @@ where
 
         log::trace!(request_id, "enqueueing two-part message to nats");
 
+        // Insert Trace Context into Headers
+        // Enables span to be created in push_endpoint before
+        // payload is parsed
+
+        let mut headers = HeaderMap::new();
+        if let Some(trace_context) = get_distributed_tracing_context() {
+            headers.insert("traceparent", trace_context.create_traceparent());
+            if let Some(tracestate) = trace_context.tracestate {
+                headers.insert("tracestate", tracestate);
+            }
+            if let Some(x_request_id) = trace_context.x_request_id {
+                headers.insert("x-request-id", x_request_id);
+            }
+            if let Some(x_dynamo_request_id) = trace_context.x_dynamo_request_id {
+                headers.insert("x-dynamo-request-id", x_dynamo_request_id);
+            }
+        }
+
         // we might need to add a timeout on this if there is no subscriber to the subject; however, I think nats
         // will handle this for us
         let _response = self
             .req_transport
-            .request(address.to_string(), buffer)
+            .request_with_headers(address.to_string(), headers, buffer)
```

Comment:
I'm fully on board with using either headers or using the header part of the TwoPartMessage object where the context id is being transferred.

Both are legit. 

You can see above this code block that we get the engine context id and pack it as part of the two part conntrol_message.

Would it be better to unify this in our control message and unpack it on our control message?

If we were to use a ZMQ push/pull socket, you don't have headers, but you have multipart messages which is kind of what our two part message maps to.

What I'm seeing here is a 3-part message.  Could we consolidate back to using our TwoPartMesage?

Reply:
I did initially put it into the control message - but it provided somehow difficult to instrument the response stream itself - and so to capture the entire request - I had to move it to the endpoint portion instead of the handler portion - I wasn't sure if we wanted to move the control message parsing .... so opted for the headers here - I think moving the parsing to the endpoint portion would be a larger change ....

I'm open to moving it or put it as a todo to consolidate -

Reply:
will add a ticket to track consolidating to the context object instead of using nats headers.